tools/online_editor/src/properties_widget.ts

```diff
@@ -53,21 +67,68 @@ export class PropertiesWidget extends Widget {
         };
     }
 
+    dispose(): void {
+        super.dispose();
+    }
+
     set_language_client(client: BaseLanguageClient | null) {
-        this.#language_client = client ?? this.#language_client;
+        if (client != null) {
+            this.#language_client = client;
+            const work = client.middleware.handleWorkDoneProgress;
+            const new_work = (
+                token: ProgressToken,
+                params:
+                    | WorkDoneProgressBegin
+                    | WorkDoneProgressReport
+                    | WorkDoneProgressEnd,
+                next: HandleWorkDoneProgressSignature,
+            ) => {
+                if (params.kind === "begin") {
+                    this.data_stale(extract_uri(params.message || ""));
+                } else if (params.kind === "end") {
+                    this.data_valid(extract_uri(params.message || ""));
+                }
+                if (work != null) {
+                    work(token, params, next);
+                }
+            };
+            this.#language_client.middleware.handleWorkDoneProgress = new_work;
+        }
     }
 
-    position_changed(uri: LspURI, version: number, position: LspPosition) {
+    data_stale(uri: string) {
+        if (uri == this.#propertiesView.current_data_uri) {
+            this.#propertiesView.show_welcome("Refreshing data");
+        }
+    }
+
+    data_valid(uri: string) {
+        if (
+            uri == this.#propertiesView.current_data_uri &&
+            this.#current_position
+        ) {
+            query_properties(this.#language_client, uri, this.#current_position)
+                .then((r: PropertyQuery) => {
+                    this.#propertiesView.set_properties(r);
+                })
+                .catch(() => {
+                    this.#propertiesView.current_data_uri = uri.toString();
+                    this.#propertiesView.show_welcome(
+                        "Data not yet available.",
+                    );
+                });
+        }
+    }
+
+    position_changed(uri: LspURI, _: number, position: LspPosition) {
+        this.#current_position = position;
         query_properties(this.#language_client, uri, position)
             .then((r: PropertyQuery) => {
                 this.#propertiesView.set_properties(r);
             })
             .catch(() => {
-                // Document has not loaded yet!
-                setTimeout(
-                    () => this.position_changed(uri, version, position),
-                    1000,
-                );
+                this.#propertiesView.current_data_uri = uri.toString();
+                this.#propertiesView.show_welcome("Data not yet available.");
```

Comment:
As a user the term “Data” is not very descriptive. Should it perhaps just be “loading…”, or is it important to tell the user about “data”?